Fixed Test, changed client to client Pointer

jrasko opened the pull request.

Andyck commented on events/event.go:
```diff
@@ -0,0 +1,49 @@
+	Time   time.Time `json:"time"`
+	Text   string    `json:"test"`
+	Source Source    `json:"source"`
+}
```
Hier benötigen wir noch die Möglichkeit benutzerdefinierte Fragmente anzulegen, zB:
```
"position": {
  "lon": 1000
  "lat": 3000
  "alt": 2000
}
```

DarkToast replied:
Ah das wusste ich nicht. Ich schau mal, wie man das in GO am geschicktesten modelliert. Evtl. was mit `map[string]interface{}`?

Andyck replied:
ja, an sowas hatte ich auch schon gedacht.

Andyck replied:
will be solved in a new branch

Andyck commented on events/event.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+type UpdateEvent struct {
+	Text string `json:"test"`
```
Hier müsste man auch benutzerdefinierte Fragmente updaten können

DarkToast replied:
Schaue ich mir im Zuge des oberen Punktes mit an.

Andyck commented on events/events.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+func (q EventQuery) QueryParams() string {
+	params := url.Values{}
```
:+1: 
url.Values{} kannte ich noch nicht

Andyck commented on events/events.go:
```diff
@@ -0,0 +1,190 @@
+	basePath string
+}
+
+func (e *events) CreateEvent(event *CreateEvent) error {
```
hier hätte ich das Event als Rückabeparameter erwartet, der im response body zurückgeliefert wird

DarkToast replied:
Beim Post kommt bei mir kein Responsebody zurück. Lediglich ein HTTP 201. Auch kein Locationheader etc:

`curl -v -X POST \
  https://tarent-gmbh.cumulocity.com/event/events \
  -H 'Authorization: Basic Yy5zY2htaWR0QHRhcmVudC5kZTpORmNCYUZod0VzN2ZDZjNEcHE=' \
  -H 'Cache-Control: no-cache' \
  -H 'Content-Type: application/json' \
  -d '{
        "type":   "TestEvent",
        "time":   "2020-06-10T07:24:29.724Z",
        "text":   "This is my test event",
        "source": { "id": "9636292" }
}'
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 52.28.191.187:443...
* TCP_NODELAY set
* Connected to tarent-gmbh.cumulocity.com (52.28.191.187) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: C=DE; ST=Nordrhein-Westfalen; L=D�sseldorf; O=Cumulocity GmbH; CN=*.cumulocity.com
*  start date: Jun  6 13:52:16 2019 GMT
*  expire date: Aug  5 10:35:38 2020 GMT
*  subjectAltName: host "tarent-gmbh.cumulocity.com" matched cert's "*.cumulocity.com"
*  issuer: C=US; ST=Arizona; L=Scottsdale; O=GoDaddy.com, Inc.; OU=http://certs.godaddy.com/repository/; CN=Go Daddy Secure Certificate Authority - G2
*  SSL certificate verify ok.
> POST /event/events HTTP/1.1
> Host: tarent-gmbh.cumulocity.com
> User-Agent: curl/7.65.3
> Accept: */*
> Authorization: Basic Yy5zY2htaWR0QHRhcmVudC5kZTpORmNCYUZod0VzN2ZDZjNEcHE=
> Cache-Control: no-cache
> Content-Type: application/json
> Content-Length: 136
> 
* upload completely sent off: 136 out of 136 bytes
* Mark bundle as not supporting multiuse
< HTTP/1.1 201 Created
< Server: openresty
< Date: Fri, 26 Jun 2020 06:17:24 GMT
< Content-Length: 0
< Connection: keep-alive
< Cache-Control: no-cache,no-store,must-revalidate
< Pragma: no-cache
< Expires: -1
< Strict-Transport-Security: max-age=31536000; includeSubDomains
< 
* Connection #0 to host tarent-gmbh.cumulocity.com left intact
`

DarkToast replied:
Beim `POST` kommt bei mir nur ein HTTP 201 - ohne Body oder Location Header:

`
curl -v -X POST \
  https://tarent-gmbh.cumulocity.com/event/events \
  -H 'Cache-Control: no-cache' \
  -H 'Content-Type: application/json' \
  -d '{
        "type":   "TestEvent",
        "time":   "2020-06-10T07:24:29.724Z",
        "text":   "This is my test event",
        "source": { "id": "9636292" }
}'^[OA
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 52.28.191.187:443...
* TCP_NODELAY set
* Connected to tarent-gmbh.cumulocity.com (52.28.191.187) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: C=DE; ST=Nordrhein-Westfalen; L=D�sseldorf; O=Cumulocity GmbH; CN=*.cumulocity.com
*  start date: Jun  6 13:52:16 2019 GMT
*  expire date: Aug  5 10:35:38 2020 GMT
*  subjectAltName: host "tarent-gmbh.cumulocity.com" matched cert's "*.cumulocity.com"
*  issuer: C=US; ST=Arizona; L=Scottsdale; O=GoDaddy.com, Inc.; OU=http://certs.godaddy.com/repository/; CN=Go Daddy Secure Certificate Authority - G2
*  SSL certificate verify ok.
> POST /event/events HTTP/1.1
> Host: tarent-gmbh.cumulocity.com
> User-Agent: curl/7.65.3
> Accept: */*
> Authorization: Basic Yy5zY2htaWR0QHRhcmVudC5kZTpORmNCYUZod0VzN2ZDZjNEcHE=
> Cache-Control: no-cache
> Content-Type: application/json
> Content-Length: 136
> 
* upload completely sent off: 136 out of 136 bytes
* Mark bundle as not supporting multiuse
< HTTP/1.1 201 Created
< Server: openresty
< Date: Fri, 26 Jun 2020 06:17:24 GMT
< Content-Length: 0
< Connection: keep-alive
< Cache-Control: no-cache,no-store,must-revalidate
< Pragma: no-cache
< Expires: -1
< Strict-Transport-Security: max-age=31536000; includeSubDomains
< 
* Connection #0 to host tarent-gmbh.cumulocity.com left intact
`

Andyck replied:
ja, das liegt wohl daran, dass accept und content-type auf "application/vnd.com.nsn.cumulocity.eventApi+json" gesetzt werden müssen. Steht relativ am Anfang: https://cumulocity.com/guides/reference/events/

DarkToast replied:
Ah ok. Schaue ich mir auch noch mal an.

DarkToast replied:
angepasst

Andyck commented on events/events.go:
```diff
@@ -0,0 +1,190 @@
+	return nil
+}
+
+func (e *events) UpdateEvent(eventId string, event *UpdateEvent) error {
```
hier könnte man auch das Event zurückgeben, das im response body enthalten ist

DarkToast replied:
Hier kommt beim `PUT` ebenfalls kein ResponseBody mit. Lediglich HTTP 204 - NoContent.

Andyck replied:
auch hier:
 das liegt wohl daran, dass accept und content-type auf "application/vnd.com.nsn.cumulocity.eventApi+json" gesetzt werden müssen.

DarkToast replied:
Schaue ich mir an.. Danke

DarkToast replied:
angepasst

Andyck commented on events/events.go:
```diff
@@ -0,0 +1,190 @@
+		return nil, nil
+	}
+
+	nextUrl, err := url.Parse(reference)
```
Wozu ist das Parsen hier nötig?

DarkToast replied:
Ich hatte in den Tests den Fall, dass die URL der Entitäten nicht funktionieren. Z.b:

Ich mache ein `GET` auf `https://tarent-gmbh.cumulocity.com/event/events`
In den Links taucht dann als `next` etc. dann aber nicht https://tarent-gmbh.cumulocity.com` sondern `https://[tenantNumber].cumulocity.com/event/events` auf. Diese Links funktionieren aber nicht mit den angegebenen Creddentials. Zumindest nicht bei mir.
Daher tausche ich hier den Host gegen den angegebenen aus.

Keine Ahnung, ob das ein Bug in der Cumulocity ist oder ob man Account einfach Mist ist. ;-)

Andyck replied:
ah, ok. Ja eigentlich sollte man dann den User t200588189/a.dyck@tarent.de statt a.dyck@tarent.de setzen. 
Ist leider etwas tricky an der Stelle.

DarkToast replied:
Ah ok. Ist natürlich eine Eingabedatengeschichte. Man kann natürlich einfach in ein Handbuch reinschreiben "Tipp den Tenant davor, Dude oder erwarte Fehler" 

Aber dann würde ich das Parsen wieder raus nehmen und direkt mit den Links arbeiten. Finde ich eh wesentlich angenehmer.

DarkToast replied:
Hmm. Ich hab es erst mal so gelassen. Kollidiert gerade mit dem http Client, der eine BaseURL mitbringt und konkateniert. 
Es funktioniert erst mal so. Vielleicht kommt mir da noch ein eleganter Geistesblitz.

Andyck commented on events/events.go:
```diff
@@ -0,0 +1,190 @@
+	return e.getCommon(fmt.Sprintf("%s?%s", e.basePath, query.QueryParams()))
+}
+
+func (e *events) NextPage(c *EventCollection) (*EventCollection, error) {
```
hmm, ich weiss nicht, ob das in der event api richtig funktioniert aber in der alarm api bekomme ich da die Referenz auf die nächste Seite obwohl es keine nächste Seite gibt. Bei der Referenz auf die vorige Seite scheint es richtig zu funktionieren.

DarkToast replied:
Stimmt. Ich baue noch einen Check ein, wenn die Response-Collection leer ist, gibts ein `nil`. Ist denke ich äquivalent zu `da ist nix mehr`

DarkToast replied:
umgesetzt

Andyck replied:
:+1:

Andyck commented on events/events.go:
```diff
@@ -0,0 +1,190 @@
+func createErrorFromResponse(responseBody []byte) error {
+	var msg map[string]interface{}
+	_ = json.Unmarshal(responseBody, &msg)
+	return errors.New(fmt.Sprintf("Request failed. Server returns error: {%s: %s}", msg["error"], msg["message"]))
```
hier würde ich generic.Error verwenden. Wird das info-Feld hier bewusst weggelassen? Ich meine, die Fehlermeldung enthält auch wichtige Infos.

DarkToast replied:
Sollten wir heute mal klären, wie wir generell in der API das Errorhandling machen. Hab jetzt oft `error` gesehen und mit dann erst nach dem PR hier auf das `generic.Error` gestoßen.

DarkToast replied:
angepasst

Andyck commented on events/client.go:
```diff
@@ -0,0 +1,64 @@
+}
+
+func (client *Client) delete(path string) ([]byte, int, error) {
+	url := client.BaseURL + path
```
`url := client.BaseURL + path` könnte man auch in request() auslagern, oder?!

DarkToast replied:
Japp. Guter Hinweis.

DarkToast replied:
umgesetzt

Andyck replied:
super

Andyck commented:
Alles in allem eine sehr gute Arbeit! Schön aufgeräumt und übersichtlich. Auch die vielen Tests sind lobenswert! Ich denke, wir könnten die pagesize als einen weiteren Parameter der EventQuery hinzufügen. Per default ist die pagesize=5. Sie darf maximal auf 2000 gesetzt werden.

DarkToast commented:
Hi Andreas,

danke für das Lob. Geh ja runter wie Butter. :-)
Die PageSize habe ich noch angepasst.

Andyck commented on events/events.go:
```diff
@@ -0,0 +1,214 @@
+func (e *events) UpdateEvent(eventId string, event *UpdateEvent) error {
+	bytes, err := json.Marshal(event)
+	if err != nil {
+		log.Printf("Error while marhalling the update event: %s", err.Error())
```
I think, we have to leave the function at this point?

Andyck commented on events/event.go:
```diff
@@ -0,0 +1,49 @@
+type CreateEvent struct {
+	Type   string    `json:"type"`
+	Time   time.Time `json:"time"`
+	Text   string    `json:"test"`
```
you mean json:"te**x**t"?

Andyck commented on events/event.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+type UpdateEvent struct {
+	Text string `json:"test"`
```
the same here: json:"te**x**t"

Andyck commented on events/event.go:
```diff
@@ -0,0 +1,49 @@
+	Type         string    `json:"type"`
+	Time         time.Time `json:"time"`
+	CreationTime time.Time `json:"creationTime"`
+	Text         string    `json:"test"`
```
...and here

Commit:
Fixed Test, changed client to client Pointer

Commit:
Fixed urls in tests.

Commit:
Deleting Wrong Query Params

Commit:
Merge remote-tracking branch 'origin/FixedTest' into FixedTest

Commit:
inventory update can be done without body now,
which is required in the test but not really necessary, so it could be better to change the test

Commit:
Improved Queryparams to something which makes sense
Fixed a bug in client.go which has been caught by my IDE (now these tests are running too)
